src/lib/autoinstall/script.rb

```diff
@@ -0,0 +1,368 @@
+require "fileutils"
+
+require "transfer/file_from_url"
+
+Yast.import "AutoinstConfig"
+
+module Y2Autoinstallation
+  # Base class representing custom user script that runs during various phases of installation.
+  class Script
+    include Yast::Logger
+    include Yast::Transfer::FileFromUrl
+    include Yast::I18n
+
+    # filename of script, used as ID
+    # @return [String]
+    attr_reader :filename
+
+    # Source code of script. Either source or location should be defined.
+    # @return [String] empty string means not defined
+    attr_reader :source
+
+    # URI for location from which download script. Either source or location should be defined.
+    # @return [String] empty string means not defined
+    attr_reader :location
+
+    # Flag if script runs in debug mode.
+    # @return [Boolean]
+    attr_reader :debug
+
+    def initialize(hash)
+      @filename = hash["filename"] || ""
+      @source = hash["source"] || ""
+      @location = hash["location"] || ""
+      @debug = !!hash["debug"]
+
+      resolve_location
+    end
+
+    def to_hash
+      {
+        "filename" => filename,
+        "source" => source,
+        "location" => location,
+        "debug" => debug
+      }
+    end
+
+    def logs_dir
+      Yast::AutoinstConfig.logs_dir
+    end
+
+    # difference from filename is that it always return non empty string
+    # @return [String]
+    def script_name
+      return filename unless filename.empty?
+
+      if location.empty?
+        self.class.type
+      else
+        url = Yast::URL.Parse(location)
+        File.basename(url["path"])
+      end
+    end
+
+    # location where script lives
+    # @return [String]
+    def script_path
+      File.join(Yast::AutoinstConfig.scripts_dir, script_name)
+    end
+
+    # Downloads or writes down script file
+    def create_script_file
+      # ensure path are available
+      Yast::SCR.Execute(Yast::Path.new(".target.mkdir"), File.dirname(script_path))
+      Yast::SCR.Execute(Yast::Path.new(".target.mkdir"), logs_dir)
+      if !location.empty?
+        url = Yast::URL.Parse(location)
+        res = get_file_from_url(
+          scheme: url["scheme"],
+          host: url["host"],
+          urlpath: url["path"],
+          localfile: localfile,
+          urltok: url,
+          destdir: Yast::AutoinstConfig.destdir
+        )
+        # TODO: exception?
+        log.error "script #{location} could not be retrieved" unless res
+      elsif !source.empty?
+        Yast::SCR.Write(Yast::Path.new(".target.string"), script_path, source)
+      else
+        log.error "Neither location or source specified for #{inspect}"
+      end
+    end
+
+  protected
+
+    # where to download file for get_file_from_url which do not switch itself chroot
+    # Just work-around for not respecting changed chroot (see bsc#897212 and bsc#889931))
+    def localfile
+      script_path
+    end
+
+  private
+
+    def resolve_location
+      return if @location.empty?
```

Comment:
np: Why not use `#location` (for consistency reasons, as you already used `#location` in `#script_name`).